luacompiler/src/bytecode/mod.rs

```diff
@@ -0,0 +1,140 @@
+pub mod instructions;
+
+use bincode::{ serialize, deserialize };
+use std::vec::Vec;
+use std::fs::File;
+use std::fmt;
+use std::io::{ Result, Write };
+use constants_map::ConstantsMap;
+
+/// A simpler representation of Lua
+#[derive(Serialize, Deserialize)]
+pub struct LuaBytecode {
+    reg_count: u8,
+    ints: Vec<i64>,
+    floats: Vec<f64>,
+    strings: Vec<String>,
+    block: Vec<u32>
+}
+
+impl LuaBytecode {
+    /// Create a new bytecode structure.
+    /// * `instrs` - the instructions of the bytecode
+    /// * `const_map` - a mapping between constants and their index in the constant table
+    /// * `reg_count` - the total number of registers used by the instructions
+    pub fn new(instrs: Vec<u32>, const_map: ConstantsMap, reg_count: u8) -> LuaBytecode {
+        LuaBytecode {
+            reg_count,
+            ints: const_map.get_ints(),
+            floats: const_map.get_floats(),
+            strings: const_map.get_strings(),
+            block: instrs
+        }
+    }
+
+    /// Create a new bytecode structure out of the given bytes.
+    /// * `bytes` - the serialized version of a LuaBytecode instance
+    /// # Panics
+    /// This panics if the given vector of bytes does not represent a LuaBytecode instance
+    pub fn new_from_bytes(bytes: Vec<u8>) -> LuaBytecode {
+        deserialize(&bytes[..]).unwrap()
+    }
+
+    /// Get the number of instructions that are part of this block.
+    pub fn instrs_len(&self) -> usize {
+        self.block.len()
+    }
+
+    /// Get the list of instructions that can be executed in order
+    /// to perform some computation.
+    pub fn get_instr(&self, index: usize) -> u32 {
+        self.block[index]
+    }
+
+    /// Get the number of registers that this bytecode uses in order to encode
+    /// instructions.
+    pub fn reg_count(&self) -> u8 {
+        self.reg_count
+    }
+
+    /// Retrieve the integer at index <i> in the constant table.
+    pub fn get_int(&self, i: u8) -> i64 {
+        self.ints[i as usize]
+    }
+
+    /// Retrieve the float at index <i> in the constant table.
+    pub fn get_float(&self, i: u8) -> f64 {
+        self.floats[i as usize]
+    }
+
+    /// Retrieve the string at index <i> in the constant table.
+    pub fn get_string(&self, i: u8) -> String {
+        self.strings[i as usize].to_string()
```

Comment:
This is, in essence, `clone`ing the string every time it's accessed. Do we want that? Could we return a `&str`?